k2/csrc/aux_labels.cc

```diff
@@ -6,14 +6,89 @@
 
 #include "k2/csrc/aux_labels.h"
 
+#include <algorithm>
 #include <numeric>
+#include <utility>
 #include <vector>
 
 #include "glog/logging.h"
 #include "k2/csrc/fsa.h"
+#include "k2/csrc/fsa_util.h"
+#include "k2/csrc/properties.h"
+
+namespace {
+
+/*
+  This function counts how many extra states we need to create for each state in
+  the input FSA when we invert an FSA. Generally, if an entering arc of state
+  `i` in the input FSA has n olabels, then we need to create n-1 extra states
+  for state `i`.
+
+  @param [in] fsa_in      Input FSA
+  @param [in] labels_in   Aux-label sequences for the input FSA
+  @param [out] expanded_states  For state `i` in `fsa_in`, we need to create
+                                extra `expaned_states[i]` states in the output
+                                inverted FSA.
+*/
+static void CountExpandedStatesNum(const k2::Fsa &fsa_in,
+                                   const k2::AuxLabels &labels_in,
+                                   std::vector<int32_t> *expanded_states) {
+  CHECK_EQ(expanded_states->size(), fsa_in.NumStates());
+  auto &states = *expanded_states;
+  for (int32_t i = 0; i != fsa_in.arcs.size(); ++i) {
+    const auto &arc = fsa_in.arcs[i];
+    int32_t pos_start = labels_in.start_pos[i];
+    int32_t pos_end = labels_in.start_pos[i + 1];
+    states[arc.dest_state] += std::max(0, pos_end - pos_start - 1);
+  }
+}
+
+/*
+  Map the state in the input FSA to state in the output inverted FSA.
+
+  @param [in] expanded_states  Output of function `CountExpandedStatesNum`
+                               which gives how many extra states we need
+                               to create for each state in the input FSA.
+  @param [out] state_map       Map state `i` in the input FSA to state
+                               `state_map[i]` in the output FSA.
+  @param [in] state_ids        state_id[i] == state_map[i-1], it will be
+                               used to number extra states created for
+                               state i.
+*/
+static void MapStates(const std::vector<int32_t> &expanded_states,
+                      std::vector<int32_t> *state_map,
+                      std::vector<int32_t> *state_ids) {
+  CHECK_EQ(state_map->size(), expanded_states.size());
+  CHECK_EQ(state_ids->size(), expanded_states.size());
+  auto &s_map = *state_map;
+  auto &s_ids = *state_ids;
+  // we suppose there's no arcs entering the start state (i.e. state id of the
+  // start state in output FSA will be 0), otherwise we may need to create a new
+  // state as the real start state.
+  CHECK_EQ(expanded_states[0], 0);
+  CHECK_EQ(s_map[0], 0);
+  CHECK_EQ(s_ids[0], 0);
+  auto num_states_in = expanded_states.size();
+  // process from the second state
+  int32_t num_states_out = 0;
+  for (auto i = 1; i != num_states_in; ++i) {
+    s_ids[i] = num_states_out;
+    // `+1` as we did not count state `i` itself in `expanded_states`
+    num_states_out += expanded_states[i] + 1;
```

Comment:
```
states[arc.dest_state] += std::max(0, pos_end - pos_start - 1);
```

If `pos_end - pos_start - 1` is `-1`, then 
```
num_states_out += expanded_states[i] + 1;
```
an extra state is added to `num_states_out`.

Maybe it should be
```
num_states_out += expanded_states[i] == 1 ? 1 : expanded_states + 1;
```